---
title: Genome-wide cfDNA fragmentation profiles comprise chromatin structure from peripheral blood cells and pancreatic cancer
---


Comparison of plasma fragmentation features to reference A/B compartments across chromosome 14. The first panel shows chromatin A/B compartments derived from pancreatic cancer tissue methylation(36). The second panel shows a median deconvoluted pancreatic cancer component based on the six samples with the highest ctDNA levels. The third panel shows the median fragmentation profile in the plasma for those samples, and the fourth panel shows the median fragmentation profile for a set of healthy plasma controls. The final panel shows chromatin A/B compartments for lymphoblast cells. Dark shading indicates regions of the genome where the two reference tracks are discordant in domain (open/closed) or magnitude. The extracted pancreatic cancer component has the greatest similarity to the pancreatic cancer reference track, and the healthy plasma has the greatest similarity to the lymphoblast reference track. 


```{r packages, message = FALSE}
library(cowplot)
library(data.table)
library(tidyverse)
library(here)
source(here("code/functions.R"))
```


```{r load_data}
sel_chr <- readLines(here("data/sel_chr.txt"))
chr_label <- paste("Chromosome", c(c(1:22), "X"))
names(chr_label) <- paste0("chr", c(c(1:22), "X"))

# Load all tracks
data <- readRDS(here("output/process_ab.Rmd/combined_bins.rds")) %>%
  group_by(source2) %>%
  mutate(newbin = row_number()) %>%
  ungroup() %>%
  setDT(.)
```


```{r}
# Select ratios / coverage
ratio_notcov <- TRUE

if (ratio_notcov) {
  sources <- c("AB Compartments PAAD",
               "Extracted Pancreatic component Ratios",
               "Median Pancreatic Plasma Ratios",
               "Median Healthy Plasma Ratios",
               "EBV transformed lymphoblast HiC")
} else {
  sources <- c("AB Compartments PAAD",
               "Extracted Pancreatic component Coverage",
               "Median Pancreatic Plasma Coverage",
               "Median Healthy Plasma Coverage",
               "EBV transformed lymphoblast HiC")
}

labels <- c("Pancreatic cancer tissue reference A/B compartments",
            "Extracted pancreatic cancer patient cfDNA component",
            "Pancreatic cancer patient cfDNA",
            "Individuals without cancer cfDNA",
            "Lymphoblastiod cell Hi-C reference A/B compartments")

# Get bins with different domains or sig difference between TCGA and Lymphoblast
clist <- chr_wrangling(data, sel_chr = sel_chr, sources, labels)

tib <- clist[["tib"]]
track_data <- clist[["track.data"]]
slevels <- levels(track_data$source2)

# Plot the tracks
b <- ggplot(track_data,
            aes(x = newbin, y = eigen, fill = color, alpha = transp)) +
  geom_bar(stat = "Identity", width = 1) +
  facet_wrap(~ source2, ncol = 1) +
  scale_x_continuous(expand  =  c(0, 0)) +
  theme_classic(base_size = 24) +
  theme(legend.position = "none",
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  coord_cartesian(xlim = c(0, 500)) +
  scale_fill_identity() +
  scale_alpha_identity() +
  xlab(chr_label[sel_chr]) +
  ylab("Fragmentation Profile")

# Plot the plasma tracks to highlight the difference
p <- tib %>%
  filter(source2 == "Non-cancer Plasma") %>%
  ggplot(aes(x = newbin, y = eigen, fill = color, alpha = transp)) +
  geom_bar(stat = "Identity",
           width = 1, color = "black") +
  geom_bar(stat = "Identity", width = 1,
           data = tib %>%
             filter(source2 == "Pancreatic Cancer Plasma"), color = "white") +
  scale_x_continuous(expand  =  c(0, 0)) +
  theme_classic(base_size = 24) +
  theme(legend.position = "none",
        axis.title.x = element_text(size = 25),
        strip.background = element_blank()) +
  scale_fill_manual(values = c("red4", "gray50")) +
  scale_alpha_identity() +
  xlab(chr_label[sel_chr]) +
  ylab("Fragmentation Profile")
```


```{r fig4, fig.width = 16, fig.height = 22, dev = c("png", "pdf")}
print(b)
```


```{r fig4-diff, fig.width = 16, fig.height = 26, dev = c("png", "pdf"), eval = FALSE}
cowplot::plot_grid(b, p, nrow = 2, rel_heights = c(2, 0.5),
                   align = "v", axis = "l")
```
